File: 2018/applelang/ObjCSolutions/Day01.m
//
//  Day01.m
//  AOCObjCTests
//
//  Created by Sam Krishna on 12/26/18.
//  Copyright © 2018 SectorMobile. All rights reserved.
//

#import <XCTest/XCTest.h>

@interface Day01 : XCTestCase
@property (nonatomic, readonly, strong) NSArray<NSNumber *> *numbers;
@end

@implementation Day01

- (NSArray<NSNumber *> *)numbers
{
    static NSMutableArray *arrayM;
    static dispatch_once_t onceToken;
    dispatch_once(&onceToken, ^{
        NSString *path = [[NSBundle bundleForClass:[self class]] pathForResource:@"day01" ofType:@"txt"];
        NSString *inputString = [NSString stringWithContentsOfFile:path encoding:NSUTF8StringEncoding error:NULL];
        arrayM = [NSMutableArray array];

        [inputString enumerateLinesUsingBlock:^(NSString * _Nonnull line, BOOL * _Nonnull stop) {
            [arrayM addObject:@(line.integerValue)];
        }];
    });

    return [arrayM copy];
}

- (void)testDay01PartA
{
    __block NSInteger sum = 0;
    [self.numbers enumerateObjectsUsingBlock:^(NSNumber * _Nonnull num, NSUInteger idx, BOOL * _Nonnull stop) {
        sum += num.integerValue;
    }];

    XCTAssertTrue(sum == 547);
}

- (void)testDay01PartB
{
    __block NSInteger frequency = 0;
    __block NSNumber *freqNumber = nil;
    NSMutableSet *frequencySet = [NSMutableSet set];

    do {
        for (NSNumber *delta in self.numbers) {
            frequency += delta.integerValue;
            NSNumber *freq = @(frequency);

            if ([frequencySet containsObject:freq]) {
                freqNumber = freq;
                break;
            }

            [frequencySet addObject:freq];
        }
    } while (!freqNumber);

    XCTAssertTrue(freqNumber.unsignedIntegerValue == 76414);
}

@end

File: 2018/applelang/SwiftSolutions/Day01.swift
//
//  Day01.swift
//  AOCSwiftTests
//
//  Created by Sam Krishna on 12/26/18.
//  Copyright © 2018 SectorMobile. All rights reserved.
//

import XCTest

class Day01: XCTestCase {

    var numbers: [Int] {
        var _numbers = type(of: self).input.components(separatedBy: "\n").map({ Int($0) ?? 0})
        _numbers.removeLast()
        return _numbers
    }

    static var input: String = {
        let corpusPath = Bundle(for: Day01.self).path(forResource: "day01", ofType: "txt")!
        let _corpus = try! String(contentsOfFile: corpusPath, encoding: String.Encoding.utf8)
        return _corpus
    }()

    func testDay01PartA() {
        let numberSum = numbers.reduce(0, { x, y in
            x + y
        })

        XCTAssertTrue(numberSum == 547)
    }

    func testDay01PartB() {
        var frequency = 0;
        var frequencySet: Set<Int> = Set()
        var shouldContinue = true

        repeat {
            for delta in numbers {
                frequency += delta

                if frequencySet.contains(frequency) {
                    shouldContinue = false
                    break
                }

                frequencySet.insert(frequency)
            }
        } while shouldContinue

        XCTAssertTrue(frequency == 76414)
    }
}
